### PR TITLE
Read only dictionaries support

### LINKED ISSUE #2794: Support readonly collection types
See #2791 

The generic collection types also implement readonly versions of themselves. 

 - `List<T> : IReadOnlyCollection<T>, IReadOnlyList<T>`
 - `Collection<T> : IReadOnlyCollection<T>`
 - `Dictionary<TKey, TValue> : IReadOnlyCollection<KeyValuePair<TKey,TValue>>, IReadOnlyDictionary<TKey,TValue>`
- `HashSet<T> : IReadOnlyCollection<T>`
- `LinkedList<T> : IReadOnlyCollection<T>`
- `Queue<T> : IReadOnlyCollection<T>`

etc

### PR DESCRIPTION
**Pull request by jbogard**
Fixes #2794

### CODE REVIEW

File: src/AutoMapper/Mappers/ReadOnlyDictionaryMapper.cs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+            var dest = Variable(dictType, "dest");
+
+            var readOnlyDictType = destExpression.Type.IsInterface
+                ? typeof(ReadOnlyDictionary<,>).MakeGenericType(dictionaryTypes)
+                : destExpression.Type;
+
+            return Block(new[] { dest }, 
+                Assign(dest, dict), 
+                Condition(NotEqual(dest, Default(dictType)), 
+                    ToType(New(readOnlyDictType.GetDeclaredConstructors().First(), dest), destExpression.Type),
```
Comment (lbargaoanu) [inline src/AutoMapper/Mappers/ReadOnlyDictionaryMapper.cs]: Maybe a more precise constructor filter? The same in CreateReadOnlyCollection.

File: src/AutoMapper/Mappers/ReadOnlyDictionaryMapper.cs [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+{
+    using static Expression;
+    using static ExpressionFactory;
+    using static CollectionMapperExpressionFactory;
+
+    public class ReadOnlyDictionaryMapper : IObjectMapper
+    {
+        public bool IsMatch(TypePair context)
+        {
+            if (!(context.SourceType.IsEnumerableType() && context.DestinationType.IsGenericType()))
```
[the remark this answers is not in the document]
Comment (jbogard) [reply src/AutoMapper/Mappers/ReadOnlyDictionaryMapper.cs]: Added one but it's slightly different than this one, as it checks for just the interface

File: src/AutoMapper/Execution/DelegateFactory.cs [hunk 3]
```diff
@@ -93,6 +92,17 @@ private static Expression CreateCollection(Type type, Type collectionType)
             return InvalidType(type, $"Cannot create an instance of interface type {type}.");
         }
 
+        private static Expression CreateReadOnlyCollection(Type type, Type collectionType)
+        {
+            var listType = collectionType.MakeGenericType(GetElementTypes(type, ElementTypeFlags.BreakKeyValuePair));
+            var ctor = listType.GetConstructors().First();
```
[the remark this answers is not in the document]
Comment (jbogard) [reply src/AutoMapper/Execution/DelegateFactory.cs]: Done.

File: src/AutoMapper/Mappers/Internal/ElementTypeHelper.cs [hunk 4]
```diff
@@ -30,6 +30,12 @@ public static class ElementTypeHelper
                 return idictionaryType.GetTypeInfo().GenericTypeArguments;
             }
 
+            var ireadonlydictionaryType = enumerableType.GetReadOnlyDictionaryType();
```
[the remark this answers is not in the document]
Comment (jbogard) [reply src/AutoMapper/Mappers/Internal/ElementTypeHelper.cs]: Done.

File: src/AutoMapper/Mappers/ReadOnlyDictionaryMapper.cs (hunk 1 above)
Comment (jbogard) [reply src/AutoMapper/Mappers/ReadOnlyDictionaryMapper.cs]: Done. Went with more precise "look for a ctor that matches exactly"

File: src/UnitTests/Mappers/ReadOnlyDictionaryMapperTests.cs [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+                    {1, 1},
+                    {2, 2},
+                    {3, 3},
+                    {4, 4},
+                }
+            };
+
+            var dest = Mapper.Map<Destination>(source);
+
+            dest.Values.Count.ShouldBe(4);
```
[the remark this answers is not in the document]
Comment (jbogard) [reply src/UnitTests/Mappers/ReadOnlyDictionaryMapperTests.cs]: Done.

**Pull request merged**